pkg/instancegroups/instancegroups.go

```diff
@@ -152,17 +149,61 @@ func (r *RollingUpdateInstanceGroup) RollingUpdate(rollingUpdateData *RollingUpd
 	settings := resolveSettings(cluster, r.CloudGroup.InstanceGroup, numInstances)
 
 	runningDrains := 0
-	maxConcurrency := settings.MaxUnavailable.IntValue()
+	maxSurge := settings.MaxSurge.IntValue()
+	if maxSurge > len(update) {
+		maxSurge = len(update)
+	}
+	maxConcurrency := maxSurge + settings.MaxUnavailable.IntValue()
 
 	if maxConcurrency == 0 {
 		klog.Infof("Rolling updates for InstanceGroup %s are disabled", r.CloudGroup.InstanceGroup.Name)
 		return nil
 	}
 
+	if r.CloudGroup.InstanceGroup.Spec.Role == api.InstanceGroupRoleMaster && maxSurge != 0 {
+		// Masters are incapable of surging because they rely on registering themselves through
```

Comment:
Not sure how chatty it would be if users don't set any values, but a warning might be great here if we're going to override user settings.

Reply:
I believe we should put on the `MaxSurge` field's documentation `Does not have any effect on instance groups with role "master".`

I could add an api validation that throws a `field.Forbidden` if `MaxSurge` is explicitly set to a non-zero value on an InstanceGroup with role "master", because that just doesn't make sense.

For the case where the user set a nonzero default `MaxSurge` on the Cluster but didn't override that at the InstanceGroup level, I believe the best user experience would be for kops to silently do the right thing. We shouldn't make the user have to explicitly override the value on each of their master InstanceGroups just to get rid of log noise.

Reply:
Ah - I missed the override point.  I agree with either/both of the suggestions, I also don't consider them a blocker to merging this.